staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go

```diff
@@ -69,24 +78,56 @@ func (e *mutatingWorkEstimator) estimate(r *http.Request) WorkEstimate {
 	// we will work on tuning the algorithm later. As a starting point we
 	// we simply assume that processing 1 event takes 1/Nth of a seat for
 	// M milliseconds and processing different events is infinitely parallelizable.
-	// We simply record the appropriate values here and rely on potential
-	// reshaping of the request if the concurrency limit for a given priority
-	// level will not allow to run request with that many seats.
-	//
+	// We simply estimate the total amount of final work for this request based
+	// on the above and rely on potential reshaping of the request if the
+	// concurrency limit for a given priority level will not allow to run
+	// request with that many seats.
+	var finalSeats uint
+	var additionalLatency time.Duration
+
 	// TODO: As described in the KEP, we should take into account that not all
 	//   events are equal and try to estimate the cost of a single event based on
 	//   some historical data about size of events.
-	var finalSeats uint
-	var additionalLatency time.Duration
+	maxSeats := uint(math.Ceil(float64(watchCount) / watchesPerSeat))
+	finalWork := SeatsTimesDuration(float64(maxSeats), eventAdditionalDuration)
 
 	// TODO: Make this unconditional after we tune the algorithm better.
 	//   Technically, there is an overhead connected to processing an event after
 	//   the request finishes even if there is a small number of watches.
 	//   However, until we tune the estimation we want to stay on the safe side
 	//   an avoid introducing additional latency for almost every single request.
 	if watchCount >= watchesPerSeat {
-		finalSeats = uint(math.Ceil(float64(watchCount) / watchesPerSeat))
-		additionalLatency = eventAdditionalDuration
+		finalSeats = maxSeats
+
+		// While processing individual events is highly parallelizable,
+		// our design/implementation have a couple limitations that would make
+		// it highly inefficient:
+		// - we reserve max(initialSeats, finalSeats) for time of executing
+		//   both phases of the request
+		// - even more importantly, when a given `wide` request is the one to
+		//   be dispatched, we are not dispatching any other request until
+		//   we accumulate enough seats to dispatch the nominated one, even
+		//   if currently unoccupied seats would allow for dispatching some
+		//   other requests in the meantime
+		// As a consequence of these, the wider the request, the more capacity
+		// will effectively be blocked and unused during dispatching and
+		// executing this request.
+		//
+		// To mitigate the impact of it, we're capping the maximum number of
+		// seats that can be assigned to a given request. Thanks to it:
+		// 1) we reduce the amount of seat-seconds that are "wasted" during
+		//    dispatching and executing initial phase of the request
+		// 2) we are not changing the finalWork estimate - just potentially
+		//    making it much longer. This is fine as long as we will not
```

Comment:
It is not "fine".  This is abandoning the correspondence between the way the actual work is scheduled --- all the notification work is launched at once --- and how APF is reserving server capacity for that work.  The hope here is that the newly relaxed relationship is good enough.

Reply:
reworded